src/mongo/db/auth/oidc/oidc_identity_providers_registry.cpp

```diff
@@ -0,0 +1,191 @@
+    getOidcIdentityProvidersRegistry(serviceContext) = std::move(registry);
+}
+
+// `_hfidps` stands for configurations of the identity providers that support human flows.
+// The correctness of the code below (see initialization list) stems from the fact that server
+// parameters validation (@see the `src/mongo/db/auth/oidc/oidc_server_parameters.cpp` file) ensures
+// that the configurations of all identity providers that support human flows are listed _before_
+// those without such support.
+OidcIdentityProvidersRegistry::OidcIdentityProvidersRegistry(
+    PeriodicRunner* periodicRunner,
+    const JWKSFetcherFactory& jwksFetcherFactory,
+    const std::vector<OidcIdentityProviderConfig>& configs)
+    : _idps(configs), _hfidps{_idps.begin(), std::ranges::find_if_not(_idps, [](const auto& idp) {
+                                  return idp.getSupportsHumanFlows();
+                              })} {
+    invariant(periodicRunner);
+
+    for (const auto& idp : _idps) {
+        // create a JWKManager instance for each issuer
+        auto res =
+            _jwkManagers.try_emplace(idp.getIssuer().toString(),
+                                     std::make_shared<crypto::JWKManager>(
+                                         jwksFetcherFactory.makeJWKSFetcher(idp.getIssuer())));
+
+        // Skip the periodic job creation if the issuer is already present in the map or the
+        // JWKSPollSecs is not configured.
+        if (!res.second || idp.getJWKSPollSecs() <= 0) {
+            continue;
+        }
+
+
+        // Create and start a periodic job which polls the JWKs from the issuer.
+        const auto period = Milliseconds{Seconds(idp.getJWKSPollSecs())};
+        auto job = [issuer = res.first->first, jwkManager = res.first->second](Client* client) {
+            Status status = jwkManager->loadKeys();
+            if (!status.isOK()) {
+                LOGV2_WARNING(29140,
+                              "Failed to load JWKs from issuer",
+                              "issuer"_attr = issuer,
+                              "error"_attr = status);
+            }
+        };
+
+        _jobs.emplace_back(periodicRunner->makeJob(
+            PeriodicRunner::PeriodicJob(fmt::format("JWKSPollJob[{}]", idp.getIssuer()),
+                                        std::move(job),
+                                        period,
+                                        /* isKillableByStepdown = */ false)));
+        _jobs.back().start();
+    }
+}
+
+boost::optional<const OidcIdentityProviderConfig&> OidcIdentityProvidersRegistry::getIdp(
+    const std::string& issuer, const std::vector<std::string>& audience) const {
+
+    // Iterate over the identity providers and check if the issuer matches the identity provider's
+    // configuration and if the audience is in the list of given audiences.
+    std::unordered_set<std::string> audienceSet{audience.begin(), audience.end()};
```

Comment:
It probably won't make a difference in the real life due to short string optimization but `std::unordered_set<std::string>` potentially does unnecessary memory allocations for string copying. Can we use `std::unordered_set<std::string_view>` instead so that its elements are always small objects that don't require memory allocation?